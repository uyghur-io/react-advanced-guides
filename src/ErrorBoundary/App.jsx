import React from 'react';

class ErrorBoundary extends React.Component {
  constructor(props) {
    super(props);
    this.state = {error: false};
  }

  static getDerivedStateFromError(error) {
    return {error: true};
  }

  componentDidCatch(error, errorInfo) {
    console.log('DEBUG: start componentDidCatch');
    console.warn(error);
    console.warn(errorInfo);
    console.log('DEBUG: end   componentDidCatch');
  }

  render() {
    if (this.state.error) {
      return <h1>Something went wrong</h1>;
    }
    return this.props.children;
  }
}

const Comp1 = props => {
  console.log('rendering inside Comp1');
  return <>
    <div>div inside Comp1</div>
    <div>props.children (inside Comp1 render()) -> {props.children}</div>
    <Comp2 {...props} />
  </>;
};

const Comp2 = props => {
  console.log('rendering inside Comp2');
  return <>
    <div>div inside Comp2</div>
    <div>props.children (inside Comp1 render()) -> {props.children}</div>
    <Comp3 {...props} />
  </>;
};

const Comp3 = class extends React.Component {

  componentDidMount() {
    console.log(`Comp3 componentDidMount -> ${this.props.n} / ${this.props.m} = ${+this.props.n / +this.props.m}`);
    if (+this.props.m === 0) {
      throw Error('Vadim: Error inside innermost component: Division by zero');
    }
  }

  render() {
    console.log('rendering inside Comp3');
    return <>
      <div>div inside Comp3</div>
      <div>props.children (inside Comp3 render()) -> {this.props.children}</div>
    </>;
  }
};

export default props => {
  console.log('rendering inside App');
  return <ErrorBoundary>
    <Comp1 n="3" m="2">children of Comp1 inside App</Comp1>
  </ErrorBoundary>
};