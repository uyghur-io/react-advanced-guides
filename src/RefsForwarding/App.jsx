import React from 'react';
import FancyButton from './FancyButton';

export default class App extends React.Component {
  buttonElement = React.createRef();

  componentDidMount() {
    console.log(
      `componentDidMount: buttonElement.current.innerHTML: ${this.buttonElement.current.innerHTML}`
    );
  }

  render() {
    return <FancyButton ref={this.buttonElement}>Ok</FancyButton>;
  }
}
