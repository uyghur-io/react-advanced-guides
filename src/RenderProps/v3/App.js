import React from 'react';
import cat from '../cat.png';

const OFFSET_X = 10;
const OFFSET_Y = 15;

class Cat extends React.Component {
  render() {
    const {x, y} = this.props.mouse;
    return (
      <img src={cat} alt="cat" style={{ position: 'absolute', width: '20px', left: x - OFFSET_X, top: y - OFFSET_Y }} />
    );
  }
}

class MouseWithCat extends React.Component {
  constructor(props) {
    super(props);
    this.handleMouseMove = this.handleMouseMove.bind(this);
    this.state = { x: OFFSET_X, y: OFFSET_Y };
  }

  handleMouseMove(event) {
    this.setState({
      x: event.clientX,
      y: event.clientY
    });
  }

  render() {
    return (
      <div style={{ height: '100vh' }} onMouseMove={this.handleMouseMove}>

        {/*
          We could just swap out the <p> for a <Cat> here ... but then
          we would need to create a separate <MouseWithSomethingElse>
          component every time we need to use it, so <MouseWithCat>
          isn't really reusable yet.
        */}
        <p>The current mouse position is ({this.state.x}, {this.state.y})</p>
        <Cat mouse={this.state} />
      </div>
    );
  }
}

class MouseTracker extends React.Component {
  render() {
    return (
      <div>
        <h1>Move the mouse around!</h1>
        <MouseWithCat />
      </div>
    );
  }
}

export default MouseTracker;