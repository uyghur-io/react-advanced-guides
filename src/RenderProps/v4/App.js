import React from 'react';
import cat from '../cat.png';

const OFFSET_X = 10;
const OFFSET_Y = 15;

class Cat extends React.Component {
  render() {
    const {x, y} = this.props.mouse;
    return (
      <img src={cat} alt="cat" style={{ position: 'absolute', width: '20px', left: x - OFFSET_X, top: y - OFFSET_Y }} />
    );
  }
}

class Mouse extends React.Component {
  constructor(props) {
    super(props);
    this.handleMouseMove = this.handleMouseMove.bind(this);
    this.state = { x: OFFSET_X, y: OFFSET_Y };
  }

  handleMouseMove(event) {
    this.setState({
      x: event.clientX,
      y: event.clientY
    });
  }

  render() {
    return (
      <div style={{ height: '100vh' }} onMouseMove={this.handleMouseMove}>
        {/*
          Instead of providing a static representation of what <Mouse> renders,
          use the `render` prop to dynamically determine what to render.
        */}
        <p>The current mouse position is ({this.state.x}, {this.state.y})</p>
        {this.props.render(this.state)}
      </div>
    );
  }
}

class MouseTracker extends React.Component {
  render() {
    return (
      <div>
        <h1>Move the mouse around!</h1>
        <Mouse render={coords => <Cat mouse={coords} />}/>
      </div>
    );
  }
}

export default MouseTracker;