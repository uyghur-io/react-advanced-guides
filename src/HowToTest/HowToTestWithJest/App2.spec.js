/**
* @jest-environment node
*/

// - the above is needed only disabled test to
// - avoid "Error: Cross origin http://localhost forbidden"
// - https://medium.com/@joenjenga/its-jest-common-problem-faced-using-jest-9905e96db8a
import React from 'react';
import {create, act} from 'react-test-renderer';
 
import App from './App';
 

describe.skip('App', () => {

  test('snapshot renders using act()', async () => {
    let component;

    await act(async () => {
      component = create(<App />);
      await new Promise(resolve => setTimeout(() => resolve(), 1000));
    });
    
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();

  });

});