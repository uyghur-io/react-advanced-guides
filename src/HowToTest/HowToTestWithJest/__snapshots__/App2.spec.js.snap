// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot renders using act() 1`] = `
<div>
  <h1>
    My Counter
  </h1>
  <div>
    <p>
      0
    </p>
  </div>
  <button
    onClick={[Function]}
    type="button"
  >
    Increment
  </button>
  <button
    onClick={[Function]}
    type="button"
  >
    Decrement
  </button>
  <h2>
    My Async Data
  </h2>
  <ul>
    <li>
      Relicensing React, Jest, Flow, and Immutable.js
    </li>
    <li>
      Build Your Own React
    </li>
    <li>
      React Native is now open source
    </li>
    <li>
      Explaining React's license
    </li>
    <li>
      React 16
    </li>
    <li>
      React Native for Android
    </li>
    <li>
      Facebook Announces React Fiber, a Rewrite of Its React Framework
    </li>
    <li>
      Painting with Code: Introducing our new open source library React Sketch.app
    </li>
    <li>
      Create React Apps with No Configuration
    </li>
    <li>
      Radical hydrogen-boron reactor leapfrogs current nuclear fusion tech?
    </li>
    <li>
      Add Reactions to Pull Requests, Issues, and Comments
    </li>
    <li>
      Vue.js vs. React
    </li>
    <li>
      React Native for Windows and Mac
    </li>
    <li>
      React v15.0
    </li>
    <li>
      Free React.js Fundamentals Course
    </li>
    <li>
      Building all of our new mobile apps using React Native
    </li>
    <li>
      Svelte 3: Rethinking Reactivity
    </li>
    <li>
      React Native for OS X
    </li>
    <li>
      Removing User Interface Complexity, or Why React is Awesome
    </li>
    <li>
      Simple React Patterns
    </li>
  </ul>
</div>
`;
